proposals/0163-lift-cpi-caller-restriction.md

```diff
@@ -0,0 +1,103 @@
+---
+simd: '0163'
+title: Lift the CPI caller restriction
+authors:
+  - Alexander Meißner
+category: Standard
+type: Core
+status: Draft
+created: 2024-07-16
+feature: TBD
+---
+
+## Summary
+
+Remove the check which forces CPI callers to have the program account of the
+callee available to them as an instruction account.
+
+## Motivation
+
+The restriction is purely historical and not necessary for the protocol or its
+implementations.
+
+Removing it would improve composability and reduce transaction building
+complexity because program accounts would no longer be passed all the way from
+transaction level instructions to the inner most CPI instructions.
+
+Additionally, not passing these program accounts is a significant reduction in
+data copied during serialization as program accounts of loader v1, v2 and v4
+contain the entire 10 MB ELF. This massively reduces the CU cost for nested CPI
+and disincentivizes programs from accessing program accounts, which could later
+allow us to remove program loading from transaction loading.
+
+### Example transaction
+
+- Accounts:
+  - Fee payer
+  - Caller Program
+  - Callee Program
+  - Token
+- Transaction-level instruction:
+  - Program Account: Caller Program
+  - Instruction Accounts:
+    - **Callee Program** (this could be removed)
+    - Token
+  - CPI:
+    - Program Account: Callee Program
+    - Instruction Accounts:
+      - Token
+
+Currently, in order to be used as a program account for a nested instruction
+the callee program must be passed as an instruction account to all outer
+instructions recursively.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+In CPI a feature gate must switch the search for the program id of the callee
+from the instruction accounts of the caller to the account list of the
+transaction.
+
+Invoking a program in CPI which was un/re/deployed in the same transaction is
+prevented by the "delay visibility" feature and thus unproblematic.
```

Comment:
It seems like this change is pretty well-contained to the `sol_invoke_signed` syscall, so maybe we should add a general spec for what that syscall does, and where exactly these changes need to occur? We don't need to explain anything Agave-specific, but there should be a protocol-level spec for `sol_invoke_signed` in this proposal.

What do you think @ripatel-fd ?

Reply:
the sol_invoke_signed syscalls are so complex I'd rather get this PR through first, nuke sol_invoke_signed_rust from  orbit, and then maybe specify sol_invoke_signed_c

Reply:
should probably rename `sol_invoke_signed_c` -> `sol_invoke_signed`

Reply:
> should probably rename `sol_invoke_signed_c` -> `sol_invoke_signed`

Sure, but I think that's beyond the scope of this SIMD. Marking resolved if all parties think we don't need to spec `sol_invoke_signed_*` here.